EIPS/eip-erc721-marketplace-extension.md

```diff
@@ -0,0 +1,323 @@
+  * @param tokenId_    : identifier of the token being listed
+  * @param price_      : the sale price of the token being listed
+  * @param tokenOwner_ : current owner of the token
+  * @param buyer_      : optional address the token is being sold too
+  */
+  function _listItem( uint256 tokenId_, uint256 price_, address tokenOwner_, address buyer_ ) internal {
+    // Update existing listing or create a new listing
+    if ( _isForSale[ tokenId_ ] ) {
+      _listings[ _listingIndex[ tokenId_ ] ].tokenprice = price_;
+      _listings[ _listingIndex[ tokenId_ ] ].to = buyer_;
+    }
+    else {
+      Listing memory _listing_ = Listing( tokenId_, price_, buyer_ );
+      _listings.push( _listing_ );
+    }
+    emit Listed( tokenId_, expiry_, tokenOwner_, price_ );
+  }
+
+  /**
+  * @dev Removes the listing for `tokenId_`.
+  * 
+  * @param tokenId_ : identifier of the token being listed
+  */
+  function _removeListing( uint256 tokenId_ ) internal {
+    uint256 _len_ = _listings.length;
+    uint256 _index_ = _listingIndex[ tokenId_ ];
+    if ( _index_ + 1 != _len_ ) {
+      _listings[ _index_ ] = _listings[ _len_ - 1 ];
+    }
+    _listings.pop();
+  }
+
+  /**
+  * @dev Processes an ether of `amount_` payment to `recipient_`.
+  * 
+  * @param amount_    : the amount to send
+  * @param recipient_ : the payment recipient
+  */
+  function _processEthPayment( uint256 amount_, address recipient_ ) internal {
+    ( boold _success_, ) = payable( recipient_ ).call{ value: amount_ }( "" );
+    require( success, "Ether Transfer Fail" );
+  }
+
+  /**
+  * @dev Transfers `tokenId_` from `fromAddress_` to `toAddress_`.
+  *
+  * This internal function can be used to implement alternative mechanisms to perform 
+  * token transfer, such as signature-based, or token burning.
+  * 
+  * @param fromAddress_ : previous owner of the token
+  * @param toAddress_   : new owner of the token
+  * @param tokenId_     : identifier of the token being transferred
+  * 
+  * Emits a {Transfer} event.
+  */
+  function _transfer( address fromAddress_, address toAddress_, uint256 tokenId_ ) internal override {
+    if ( _isForSale[ tokenId_ ] ) {
+      _removeListing( tokenId_ );
+    }
+    super._transfer( fromAddress_, toAddress_, tokenId_ );
+  }
+
+  // PUBLIC
+  function listItem( uint256 tokenId, uint256 price, address to ) external {
+  /**
+  * @notice Create or update a listing for `tokenId_`.
+  * 
+  * Note: Setting `buyer_` to the NULL address will create a public listing.
+  * 
+  * @param tokenId_ : identifier of the token being listed
+  * @param price_   : the sale price of the token being listed
+  * @param buyer_   : optional address the token is being sold too
+  */
+  function listItem( uint256 tokenId_, uint256 price_, address buyer_ ) external {
+    address _tokenOwner_ = ownerOf( tokenId_ );
+    require( _tokenOwner_ == ownerOf( tokenId_ ), "Not token owner" );
+
+    _createListing( tokenId_, price_, _tokenOwner_, buyer_ );
+  }
```

Comment:
```suggestion
  function listItem( uint256 tokenId_, uint256 price_, address buyer_ ) external {
    address _tokenOwner_ = ownerOf( tokenId_ );
    require( _tokenOwner_ == ownerOf( tokenId_ ), "Not token owner" );

    _createListing( tokenId_, price_, _tokenOwner_, buyer_ );
  }
```
Designated `buyer` does not seem to be used very often. If a particular buyer and seller privately agree on a trading price, they have multiple ways to complete the trading